resources/aws.go

```diff
@@ -0,0 +1,334 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	errwrap "github.com/pkg/errors"
+)
+
+func init() {
+	RegisterResource("aws", func() Res { return &AwsRes{} })
+}
+
+// AwsRes is an aws resource.
+type AwsRes struct {
+	BaseRes `yaml:",inline"`
+	State   string `yaml:"state"` // state: running, stopped, terminated
+	Region  string `yaml:"region"`
+	Type    string `yaml:"type"`
+	ImageID string `yaml:"imageid"`
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *AwsRes) Default() Res {
+	return &AwsRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *AwsRes) Validate() error {
+	if obj.State != "running" && obj.State != "stopped" && obj.State != "terminated" {
+		return fmt.Errorf("State must be 'running', 'stopped' or 'terminated'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *AwsRes) Init() error {
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *AwsRes) Watch() error {
+	send := false
+	var exit *error
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	}))
+	svc := ec2.New(sess)
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.GetName())},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending"), aws.String("stopped"), aws.String("stopping")},
+			},
+		},
+	}
+	if err := obj.Running(); err != nil {
+		return err
+	}
+	awschan := make(chan string)
+	go func() {
+		for {
+			diOutput, err := svc.DescribeInstances(&diInput)
+			if err != nil {
+				errwrap.Wrapf(err, "error describing instances")
+			}
+			var waitInput ec2.DescribeInstancesInput
+			if len(diOutput.Reservations) == 1 {
+				waitInput = ec2.DescribeInstancesInput{
+					InstanceIds: []*string{diOutput.Reservations[0].Instances[0].InstanceId},
+				}
+			} else {
+				waitInput = diInput
+			}
+
+			if obj.State == "running" && len(diOutput.Reservations) == 1 {
+				svc.WaitUntilInstanceStopped(&waitInput)
+				if *diOutput.Reservations[0].Instances[0].State.Name != "running" {
+					awschan <- "stopped"
+					svc.WaitUntilInstanceRunning(&waitInput)
+				}
+			}
+			if obj.State == "stopped" && len(diOutput.Reservations) == 1 {
+				svc.WaitUntilInstanceRunning(&waitInput)
+				if *diOutput.Reservations[0].Instances[0].State.Name != "stopped" {
+					awschan <- "running"
```

Comment:
what if we close while being blocked here? we might need a close channel signal, follow the ``closeChan` pattern for an example, see: https://github.com/purpleidea/mgmt/pull/224/files#diff-e546bf8ca2594db1f20ccb778896dfb6R98